### PR TITLE
[IMP] web: Add widget signature on binary field

### PR DESCRIPTION
**Pull request by jbm-odoo**
Add a new widget for binary fields. It open a dialog box and the user can sign manually,
or an signature can be draw automatically or he can upload a picture of his signature.

There are 3 options on the widget:
A 'full_name' option that defines which char field will be used as 'full name' to automatically generate the signature, the field must be present in the view (either in invisible). Example: "{'full_name': 'display_name'}". If this option is set, we can use the auto sign function.
A 'default_font' option to set the font to use for the 'auto' signature. If this option is set, we should not show the font selector.
A 'size' option with the length and/or the width. Example: "{'size': [200,'']}" or "{'size': [200,400]}".

id=1929011

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/web/static/src/js/fields/basic_fields.js [hunk 1]
```diff
@@ -1581,6 +1582,442 @@ var FieldBinaryImage = AbstractFieldBinary.extend({
     },
 });
 
+var SignatureDialog = Dialog.extend({
```
Comment (VincentSchippefilt) [inline addons/web/static/src/js/fields/basic_fields.js]: this is not the correct place to define this dialog, in fact the entire definition of the binarySignature widget should be in its own file in the fields folder

File: addons/web/static/src/js/fields/basic_fields.js [hunk 2]
```diff
@@ -1581,6 +1582,442 @@ var FieldBinaryImage = AbstractFieldBinary.extend({
...
+        'click a.o_sign_mode': '_onClickChangeSignatureMode',
+        'click .o_sign_select_style': '_onClickOpenFont',
+        'click .o_sign_font_dialog a': '_onClickSelectFont',
+        'click .o_sign_font_dialog a, .o_sign_signature': '_onClickHideFontDialog',
+        'click .o_sign_clean': '_onClickCleanSignature',
+        'input .o_sign_load': '_onLoadSignature',
+    },
+    /**
+     * @override
+     * @param {Object} fields
```
Comment (VincentSchippefilt) [inline addons/web/static/src/js/fields/basic_fields.js]: this @param is wrong

File: addons/web/static/src/js/fields/basic_fields.js [hunk 3]
```diff
@@ -1581,6 +1582,442 @@ var FieldBinaryImage = AbstractFieldBinary.extend({
...
+                'background-color': '#FFF',
+                'color': '#000',
+                'lineWidth': 2,
+                'width': width,
+                'height': height
+            });
+
+            self.emptySignature = self.$signatureField.jSignature("getData");
+
+            self.$modeButtons.filter('.btn-primary').click();
```
Comment (VincentSchippefilt) [inline addons/web/static/src/js/fields/basic_fields.js]: it is weird to me that you have to click a button of the dialog when you open it. You can probably set its state correctly without interacting with the DOM by calling the method the click would have called

File: addons/web/static/src/js/fields/basic_fields.js [hunk 4]
```diff
@@ -1581,6 +1582,442 @@ var FieldBinaryImage = AbstractFieldBinary.extend({
...
+     * @private
+     * Render and scale the image in the canvas.
+     * @param {string} imgSrc image (either a generated signature or a uploaded image)
+     */
+    _printImage: function (imgSrc) {
+        var self = this;
+
+        var image = new Image;
+        image.onload = function () {
+            var width = 0, height = 0;
```
Comment (VincentSchippefilt) [inline addons/web/static/src/js/fields/basic_fields.js]: the usual style in Odoo is to declare each variable on its own line

File: addons/web/static/src/js/fields/basic_fields.js [hunk 5]
```diff
@@ -1581,6 +1582,442 @@ var FieldBinaryImage = AbstractFieldBinary.extend({
...
+        this.mode = $(e.target).data('mode');
+
+        if (SignatureDialog.fonts.length > 1) {
+            this.$selectStyleButton.toggle(this.mode === 'auto');
+        }
+        this.$clearButton.toggle(this.mode === 'draw');
+        this.$loadButton.toggle(this.mode === 'load');
+
+        if (this.mode === 'load') {
+            this.$loadButton.click();
```
Comment (VincentSchippefilt) [inline addons/web/static/src/js/fields/basic_fields.js]: I would prefer that you do not programmatically click on buttons, but call the method that the button calls instead,if possible.

File: addons/web/static/src/js/fields/basic_fields.js [hunk 6]
```diff
@@ -1581,6 +1582,442 @@ var FieldBinaryImage = AbstractFieldBinary.extend({
...
+
+    /**
+     * @private
+     * Open selector with different fonts for automatic generation
+     */
+    _onClickOpenFont: function () {
+        var self = this;
+        this.$fontDialog.find('a').empty().append($('<div/>'));
+        this.$fontDialog.show().animate({'width': self.$fontDialog.find('a').first().height() * self.signatureRatio * 1.25}, 500, function () {
+            self.$fontDialog.find('a').each(function (i, el) {
```
Comment (VincentSchippefilt) [inline addons/web/static/src/js/fields/basic_fields.js]: Note that this function will only be called after the animation is completed, so after 500 ms (and not during). Is it the intention ?

File: addons/web/static/src/js/fields/basic_fields.js [hunk 7]
```diff
@@ -1581,6 +1582,442 @@ var FieldBinaryImage = AbstractFieldBinary.extend({
...
+     * @private
+     */
+    _onClickSignature: function () {
+        if (this.mode === 'edit') {
+            var self = this;
+
+            if (this.nodeOptions.full_name) {
+                var search = "[name='" + this.nodeOptions.full_name + "']";
+                // If full_name correspond to field in read-only mode
+                var signName = $(search).text();
```
Comment (VincentSchippefilt) [inline addons/web/static/src/js/fields/basic_fields.js]: if you do not store $(search) in a variable, every time you call $(search) there is a lookup on the dom.
I would replace the start of the function by
`var $search = $("[name=.....]");`
then use $search

File: addons/web/static/src/scss/fields.scss [hunk 8]
```diff
@@ -411,6 +411,18 @@
         }
     }
 
+    // Signature
+    .o_signature {
+        outline: 1px solid lightgrey;
```
Comment (VincentSchippefilt) [inline addons/web/static/src/scss/fields.scss]: try to not hardcode colors in the styles, make them depend on an existing variable

File: addons/web/static/src/scss/fields.scss [hunk 9]
```diff
@@ -446,6 +458,53 @@
 
 }
 
+.o_field_invalid {
+    .o_signature {
+        outline: 3px solid #DC6965;
```
Comment (VincentSchippefilt) [inline addons/web/static/src/scss/fields.scss]: try to not hardcode colors in the styles, make them depend on an existing variable

File: addons/web/static/src/scss/fields.scss [hunk 10]
```diff
@@ -446,6 +458,53 @@
...
+        cursor: pointer;
+    }
+}
+.o_form_editable {
+    .o_signature:hover {
+        outline: 3px solid #00A09D;
```
Comment (VincentSchippefilt) [inline addons/web/static/src/scss/fields.scss]: try to not hardcode colors in the styles, make them depend on an existing variable

File: addons/web/static/src/scss/fields.scss [hunk 11]
```diff
@@ -446,6 +458,53 @@
...
+                white-space:nowrap;
+            }
+        }
+    }
+}
+
+.o_sign_signature_draw .o_sign_signature {
+    padding: 0;
+    > canvas.jSignature {
+        outline: #ced4da 1px solid;
```
Comment (VincentSchippefilt) [inline addons/web/static/src/scss/fields.scss]: try to not hardcode colors in the styles, make them depend on an existing variable

File: addons/web/static/tests/fields/basic_fields_tests.js [hunk 12]
```diff
@@ -1841,6 +1842,72 @@ QUnit.module('basic_fields', {
...
+        testUtils.form.clickEdit(form);
+
+        testUtils.dom.click(form.$('.o_field_widget[name="signature"] > .o_signature'));
+
+        assert.strictEqual($('.o_sign_mode_draw:visible').length, 1, "there shoud have a button to draw the signature");
+        assert.strictEqual($('.o_sign_mode_auto:visible').length, 1, "there shoud have a button to auto draw the signature");
+        assert.strictEqual($('.o_sign_mode_load:visible').length, 1, "there shoud have a button to load the signature");
+
+        assert.containsOnce($, '.o_sign_font_dialog', "there shoud have a panel with fonts");
+        assert.ok($('.o_sign_font_selection > a').length > 2, "should have more than 2 fonts");
```
Comment (VincentSchippefilt) [inline addons/web/static/tests/fields/basic_fields_tests.js]: this is a bad assert: 'more than 2' is not specific, you control exactly how many to you can be specific, and use assert.containsN with that specific number

**Comment by VincentSchippefilt** [comment]
I have a few general questions about your PR:

1. do we actually use all the files we add (the font in all formats) and in which condition do we use which ?
1. You have to rework the commit messages and the commits themselves, they can probably be squashed, they have to be named correctly and a correct description of what is implemented is welcome
1. Thank you for writing unit tests, that is nice! 👍

File: addons/web/static/src/js/fields/signature.js [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+    "use strict";
+
+    var core = require('web.core');
+    var session = require('web.session');
+    var utils = require('web.utils');
+    var field_utils = require('web.field_utils');
+    var Dialog = require('web.Dialog');
+    var AbstractFieldBinary = require('web.basic_fields').AbstractFieldBinary;
+    var registry = require('web.field_registry');
+    var NameAndSignature = require('web.name_and_signature').NameAndSignature;
```
Comment (VincentSchippefilt) [inline addons/web/static/src/js/fields/signature.js]: alphabetic sort

File: addons/web/static/src/js/fields/signature.js [hunk 14]
```diff
@@ -0,0 +1,296 @@
...
+    custom_events: {
+        'signature_changed': '_onChangeSignature',
+    },
+
+    //----------------------------------------------------------------------
+    // Public
+    //----------------------------------------------------------------------
+
+    /**
+     * Allows options.
```
Comment (VincentSchippefilt) [inline addons/web/static/src/js/fields/signature.js]: ?

File: addons/web/static/src/js/fields/signature.js [hunk 15]
```diff
@@ -0,0 +1,296 @@
...
+            options.buttons = [];
+            options.buttons.push({text: _t("Adopt and Sign"), classes: "btn-primary", disabled: true, click: function (e) {
+                this.confirmFunction();
+            }});
+            options.buttons.push({text: _t("Cancel"), close: true});
+        }
+
+        this._super(parent, options);
+
+        this.confirmFunction = function () {};
```
Comment (VincentSchippefilt) [inline addons/web/static/src/js/fields/signature.js]: declare confirmFunction the same way you declare init

File: addons/web/static/src/js/fields/signature.js [hunk 16]
```diff
@@ -0,0 +1,296 @@
...
+     * Inserts the name and signature.
+     *
+     * @override
+     */
+    start: function () {
+        this.$primaryButton = this.$footer.find('.btn-primary');
+        this.nameAndSignature.replace(this.$('.o_web_sign_name_and_signature'));
+        return this._super.apply(this, arguments);
+    },
+
```
Comment (VincentSchippefilt) [inline addons/web/static/src/js/fields/signature.js]: //----------------------------------------------------------------------
    // Public
    //----------------------------------------------------------------------

Comment (VincentSchippefilt) [inline addons/web/static/src/js/fields/signature.js]: ! make private methods private (starts with _)

File: addons/web/static/src/js/fields/signature.js [hunk 17]
```diff
@@ -0,0 +1,296 @@
...
+            }
+            var signDialog = new SignatureDialog(self, {nameAndSignatureOptions: nameAndSignatureOptions});
+
+            signDialog.open().onConfirm(function () {
+                if (this.getSignatureImage() !== signDialog.emptySignature) {
+                    var data = this.getSignatureImage()[1];
+                    var type = this.getSignatureImage()[0].split('/')[1];
+                    self.on_file_uploaded(data.length, this.getName()
+                    , type, data);
+                }
```
Comment (VincentSchippefilt) [inline addons/web/static/src/js/fields/signature.js]: should be in signature dialog

File: addons/web/static/src/js/fields/signature.js [hunk 18]
```diff
@@ -0,0 +1,296 @@
...
+    //--------------------------------------------------------------------------
+
+    /**
+     * If the view is in edit mode, open dialog to sign.
+     *
+     * @private
+     */
+    _onClickSignature: function () {
+        if (this.mode === 'edit') {
+            var self = this;
```
Comment (VincentSchippefilt) [inline addons/web/static/src/js/fields/signature.js]: should be the first line of the function

File: addons/web/static/src/js/fields/signature.js [hunk 19]
```diff
@@ -0,0 +1,296 @@
...
+     */
+    start: function () {
+        this.$primaryButton = this.$footer.find('.btn-primary');
+        this.nameAndSignature.replace(this.$('.o_web_sign_name_and_signature'));
+        return this._super.apply(this, arguments);
+    },
+
+    onConfirm: function (fct) {
+        this.confirmFunction = fct;
+    },
```
Comment (VincentSchippefilt) [inline addons/web/static/src/js/fields/signature.js]: probably not usefull

File: addons/web/static/src/js/fields/signature.js [hunk 20]
```diff
@@ -0,0 +1,296 @@
...
+    /**
+     * Inserts the name and signature.
+     *
+     * @override
+     */
+    start: function () {
+        this.$primaryButton = this.$footer.find('.btn-primary');
+        this.nameAndSignature.replace(this.$('.o_web_sign_name_and_signature'));
+        return this._super.apply(this, arguments);
```
Comment (VincentSchippefilt) [inline addons/web/static/src/js/fields/signature.js]: `return this._super.apply(this, arguments).then(function() {
               return this.nameAnd....replace(...)
 }`

File: addons/web/static/src/js/fields/signature.js [hunk 21]
```diff
@@ -0,0 +1,296 @@
...
+    _onChangeSignature: function () {
+        var isEmpty = this.nameAndSignature.isSignatureEmpty();
+        this.$primaryButton.prop('disabled', isEmpty);
+    },
+});
+
+var FieldBinarySignature = AbstractFieldBinary.extend({
+    fieldDependencies: _.extend({}, AbstractFieldBinary.prototype.fieldDependencies, {
+        __last_update: {type: 'datetime'},
+    }),
```
Comment (VincentSchippefilt) [inline addons/web/static/src/js/fields/signature.js]: what is this ?

File: addons/web/static/src/js/fields/signature.js [hunk 22]
```diff
@@ -0,0 +1,296 @@
...
+
+    /**
+     * Renders an empty signature or the saved signature. Both must have the same size.
+     *
+     * @override
+     * @private
+     */
+
+    _render: function () {
+        var signatureRatio = 3;
```
Comment (VincentSchippefilt) [inline addons/web/static/src/js/fields/signature.js]: rename me to displayRatio or something clearer

File: addons/web/static/src/js/fields/signature.js [hunk 23]
```diff
@@ -0,0 +1,296 @@
...
+        var self = this;
```
Comment (VincentSchippefilt) [inline addons/web/static/src/js/fields/signature.js]: first line of the function

File: addons/web/static/src/js/widgets/name_and_signature.js [hunk 24]
```diff
@@ -257,7 +278,7 @@ var NameAndSignature = Widget.extend({
         this.$autoButton.toggleClass('active', this.signMode === 'auto');
         this.$loadButton.toggleClass('active', this.signMode === 'load');
 
-        this.$signatureField.jSignature(this.signMode === 'draw' ? 'enable' : 'disable');
+//         this.$signatureField.jSignature(this.signMode === 'draw' ? 'enable' : 'disable');
```
Comment (VincentSchippefilt) [inline addons/web/static/src/js/widgets/name_and_signature.js]: remove this

File: addons/web/static/src/js/widgets/name_and_signature.js [hunk 25]
```diff
@@ -313,7 +334,7 @@ var NameAndSignature = Widget.extend({
      * @returns {string} cleaned name
      */
     _getCleanedName: function () {
-        var text = this.getName().replace(/[^[\w\u00E0-\u00FC]-'" ]/g, '');
+        var text = this.getName().replace(/[^\w\u00E0-\u00FC-'" ]/g, '');
```
Comment (VincentSchippefilt) [inline addons/web/static/src/js/widgets/name_and_signature.js]: declare regex in a var, give it a nice name please

File: addons/web/static/src/js/widgets/name_and_signature.js [hunk 26]
```diff
@@ -565,6 +586,9 @@ var NameAndSignature = Widget.extend({
      */
```
Comment (VincentSchippefilt) [inline addons/web/static/src/js/widgets/name_and_signature.js]: missing  @return bool|undefined

**Comment by RomainLibert** [comment]
@robodoo r+ rebase-ff

**Pull request closed without merging**